marlowe-playground-client/src/Marlowe/Holes.purs

```diff
@@ -261,32 +265,77 @@ constructMarloweType constructorName (MarloweHole { range: { startColumn, startL
+  --   * See if this belongs to LinterMarker rather than Marlowe.Holes, as it's relying on the textual
+  --     representation of the marker and a Range location.
+  --   * See if we can Modify the parser so that a contract starting with parentesis is valid
+  --   * Or see if we have another way to detect the first hole that does not depend on being the first
+  --     character
+  parens (Range { startColumn: 1, startLineNumber: 1, endColumn: _, endLineNumber: _ }) text = text
 
-  parens _ _ text = "(" <> text <> ")"
+  parens _ text = "(" <> text <> ")"
 
-type Range
-  = { startLineNumber :: Int
-    , startColumn :: Int
-    , endLineNumber :: Int
-    , endColumn :: Int
-    }
+data Location
+  = Range IRange
+  | BlockId String
+  -- CHECK THIS: Before this refactor, the Range had an implicit Semiring constraint in multiple places, for
+  -- example mkDefaultTermWrapper. The idea was mainly to use a `zero`/`mempty` location. But having a semiring
+  -- is probably not what we want for Range or BlockId. For the moment I'm adding a separate constructor
+  -- to the location, but perhaps we want to modify the Term constructor to receive a (Maybe Location) instead.
+  | NoLocation
+
+derive instance eqLocation :: Eq Location
+
+derive instance genericSession :: Generic Location _
+
+instance encodeSession :: Encode Location where
+  encode value = genericEncode defaultOptions value
+
+instance decodeSession :: Decode Location where
+  decode value = genericDecode defaultOptions value
 
 -- We need to compare the fields in the correct order
-compareRange :: Range -> Range -> Ordering
+compareRange :: IRange -> IRange -> Ordering
 compareRange =
   compare `on` _.startColumn
     <> compare `on` _.startLineNumber
     <> compare `on` _.endLineNumber
     <> compare `on` _.endColumn
 
+-- FIXME: We should check the need for the different Ord's instances we have here.
+--        I've traced that all the Term's instances are required because the Linter has a generic
+--        compare on the WarningDetail, which is only used as the fallback case of ordWarning.
+--        The other use of ordering is in MarloweHole, to be able to insert them in the `Holes` Set.
+--        But that is currently using compareLocation and not the Location Ord instance. And it doesn't
+--        necesarely need to be based on location, is just a constraint for Set.
+instance ordLocation :: Ord Location where
+  compare = compareLocation
+
+compareLocation :: Location -> Location -> Ordering
+compareLocation (Range a) (Range b) = compareRange a b
+
+-- TODO: Comparing on String does not make too much sense for ordering blocks id's. Once
+--       we refactor Blockly to produce Term's directly, we can see if it makes sense to store
+--       the id and the depth and compare on depth.
+compareLocation (BlockId a) (BlockId b) = compare a b
+
+compareLocation NoLocation _ = LT
+
+compareLocation _ NoLocation = GT
+
+compareLocation _ _ = EQ
```

Comment:
I needed to add the last `_ _` because I could be comparing Range to BlocklyId and vice-versa 😓

Reply:
Sorry, by "this" I meant the whole `compareLocation` function, it says `compare = compareLocation`. So why do we need `compareLocation` at all? We can just rename `compareLocation -> compare`.

But since you mention it, maybe anything other than `EQ` would be better, because they are clearly not EQ if they are different types. Even though I guess it is kind of arbitrary comparing them, ofc, we could through an exception if we really don't want that to happen. Otherwise maybe we can just say `Range` is smaller then `BlockId`, for example.

Reply:
I take back the first paragraph (because of the other comment)

Reply:
In last commit I ended up adding an unsafeThrow if we are trying to compare BlockId and Range